File: solutions/scala/queen-attack/1/src/main/scala/QueenAttack.scala
import scala.util.Try

case class Queen(val row: Int, val col: Int)

object Queen {
  def create(row: Int, col: Int) =
    Try {
      require((row <= 7 && row >= 0) && (col <= 7 && col >= 0))
      Queen(row, col)
    }.toOption
}

object QueenAttack {
  def canAttack(one: Queen, two: Queen): Boolean = {
    !(one.row != two.row && one.col != two.col &&
      math.abs(one.row - two.row) != math.abs(one.col - two.col))
  }
}

File: solutions/fsharp/queen-attack/1/QueenAttack.fs
module QueenAttack

let create (row:int, col:int) = 
    row >= 0 && row <=7 && col >=0 && col <= 7
 
let canAttack (row1:int,col1:int) (row2:int,col2:int) = 
    let deltaR = row1 - row2
    let deltaC = col1 - col2
    deltaC = 0 || deltaR = 0 || abs(deltaC) = abs(deltaR)